pilot/cmd/istioctl/controlplane/cmd.go

```diff
@@ -0,0 +1,149 @@
+		RunE: func(c *cobra.Command, args []string) error {
+			if err := install.setFeatures(*features); err != nil {
+				return err
+			}
+			install.Namespace = *istioNamespaceFlag
+			switch strings.ToLower(out) {
+			case "helm":
+				_, err := fmt.Fprint(os.Stdout, valuesFromInstallation(install))
+				return err
+			case "yaml":
+				rendered, err := yamlFromInstallation(install, helmChartLocation)
+				if err != nil {
+					return err
+				}
+				_, err = fmt.Fprint(os.Stdout, rendered)
+				return err
+			default:
+				return fmt.Errorf("unsupported output %q", out)
+			}
+		},
+	}
+
+	features = cmd.PersistentFlags().StringArrayP("features", "f", []string{},
+		`List of Istio features to enable. Accepts any combination of "mtls", "telemetry", "routing", "ingress", "policy", "initializer".`)
+	cmd.PersistentFlags().StringVar(&install.Hub, "hub", "gcr.io/istio-testing", "The container registry to pull Istio images from")
+	cmd.PersistentFlags().StringVar(&install.MixerTag, "mixer-tag", "latest", "The tag to use to pull the `mixer` container")
+	cmd.PersistentFlags().StringVar(&install.PilotTag, "pilot-tag", "latest", "The tag to use to pull the `pilot-discovery` container")
+	cmd.PersistentFlags().StringVar(&install.CaTag, "ca-tag", "latest", "The tag to use to pull the `ca` container")
+	cmd.PersistentFlags().StringVar(&install.ProxyTag, "proxy-tag", "latest", "The tag to use to pull the `proxy` container")
+	cmd.PersistentFlags().BoolVar(&install.Debug, "debug", false, "If true, uses debug images instead of release images")
+	cmd.PersistentFlags().Uint16Var(&install.NodePort, "ingress-node-port", 0,
+		"If provided, Istio ingress proxies will run as a NodePort service mapped to the port provided by this flag. "+
+			"Note that this flag is ignored unless the \"ingress\" feature flag is provided too.")
+	cmd.PersistentFlags().StringVarP(&out, "out", "o", "helm", `Output format. Acceptable values are:
+					"helm": produces contents of values.yaml
+					"yaml": produces Kubernetes deployments`)
+
+	// TODO: figure out how we want to package up the charts with the binary to make this easy
+	cmd.PersistentFlags().StringVar(&helmChartLocation, "helm-chart-dir", ".",
+		"The directory to find the helm charts used to render Istio deployments. -o yaml uses these to render the helm chart locally.")
+
+	_ = cmd.PersistentFlags().MarkHidden("hub")
+	_ = cmd.PersistentFlags().MarkHidden("mixer-tag")
+	_ = cmd.PersistentFlags().MarkHidden("pilot-tag")
+	_ = cmd.PersistentFlags().MarkHidden("ca-tag")
+	_ = cmd.PersistentFlags().MarkHidden("proxy-tag")
+	return cmd
+}
+
+type installation struct {
+	Mixer       bool
+	Pilot       bool
+	Ca          bool
+	Ingress     bool
+	Initializer bool
+
+	Namespace string
+	// todo: support hub per component
+	Hub      string
+	MixerTag string
+	PilotTag string
+	CaTag    string
+	ProxyTag string
+	Debug    bool
+	NodePort uint16
+}
+
+func defaultInstall() *installation {
+	return &installation{
```

Comment:
We should also set default values of other fields here instead of relying of flag passing to do that.

Reply:
Flags already provide a default value, I don't really see a reason for us to specify a default twice. If we want to extract this from the command in the future and decouple the two, then sure we can define defaults in the struct then.

Reply:
I am thinking about the case of reading the installation config from some files instead of flags. We can do it later though.

Reply:
Added support for reading in the `values.yaml` file directly based on talking with Costin and Louis. In that case we basically skip the "installation" step. IMO it doesn't make sense to round trip values.yaml -> installation struct -> values.yaml -> rendered deployments instead of just going directly from values.yaml -> rendered deployments.

Longer term we might want to do that full round trip so the tool is more extensible, but that means coupling it pretty tightly to the values.yaml, since we'd need to read info out of it.